File: src/main/java/de/tum/www1/orion/util/registry/OrionProjectRegistryStateService.java
package de.tum.www1.orion.util.registry;

import com.intellij.openapi.application.ActionsKt;
import com.intellij.openapi.components.*;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.projectRoots.ProjectJdkTable;
import com.intellij.openapi.projectRoots.Sdk;
import com.intellij.openapi.roots.ProjectRootManager;
import com.intellij.openapi.vfs.VfsUtil;
import com.jetbrains.python.sdk.PythonSdkType;
import de.tum.www1.orion.dto.RepositoryType;
import de.tum.www1.orion.enumeration.ExerciseView;
import de.tum.www1.orion.enumeration.ProgrammingLanguage;
import de.tum.www1.orion.util.JsonUtilsKt;
import de.tum.www1.orion.util.OrionFileUtils;
import de.tum.www1.orion.util.UtilsKt;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

@State(name = "orionRegistry", storages = @Storage(StoragePathMacros.WORKSPACE_FILE))
public class OrionProjectRegistryStateService implements PersistentStateComponent<OrionProjectRegistryStateService.State> {
    private static final Logger log = Logger.getInstance(OrionProjectRegistryStateService.class);
    private State myState;
    private Project myProject;

    public OrionProjectRegistryStateService(Project project) {
        this.myProject = project;
    }

    public static OrionProjectRegistryStateService getInstance(@NotNull Project project) {
        return ServiceManager.getService(project, OrionProjectRegistryStateService.class);
    }

    public static class State {
        public long courseId;
        public String courseTitle;
        public long exerciseId;
        public String exerciseTitle;
        public RepositoryType selectedRepository;
        public ExerciseView view;
        public ProgrammingLanguage language;
    }

    @Nullable
    @Override
    public State getState() {
        return myState;
    }

    @Override
    public void loadState(@NotNull State state) {
        myState = state;
    }

    public void importIfPending() {
        final var pendingImportFile = VfsUtil.findRelativeFile(OrionFileUtils.INSTANCE.getRoot(myProject), ".artemisExercise.json");
        if (pendingImportFile != null) {
            try {
                final var imported = JsonUtilsKt.mapper().readValue(pendingImportFile.getInputStream(), ImportedExercise.class);
                myState = new State();
                myState.courseId = imported.getCourseId();
                myState.exerciseId = imported.getExerciseId();
                myState.courseTitle = imported.getCourseTitle();
                myState.exerciseTitle = imported.getExerciseTitle();
                myState.language = imported.getLanguage();
                myState.view = imported.getView();
                if (myState.view == ExerciseView.INSTRUCTOR) {
                    guessProjectSdk();
                    myState.selectedRepository = RepositoryType.TEST;  // init
                }

                ActionsKt.runWriteAction(() -> {
                    try {
                        pendingImportFile.delete(this);
                    } catch (IOException e) {
                        log.error(e.getMessage(),  e);
                    }

                    return null;
                });
            } catch (IOException e) {
                log.error(e.getMessage(), e);
            }
        }
    }

    private void guessProjectSdk() {
        final List<Sdk> availableSdks;
        switch (myState.language) {
            case JAVA:
                availableSdks = List.of(ProjectJdkTable.getInstance().getAllJdks());
                break;
            case PYTHON:
                availableSdks = PythonSdkType.getAllSdks();
                break;
            default:
                throw new IllegalArgumentException("Programming language " + myState.language + " is not supported yet!");
        }

        final var bestFit = availableSdks.stream().max(Comparator.comparing(sdk -> Objects.requireNonNull(sdk.getVersionString())));
        ActionsKt.runWriteAction(UtilsKt.ktLambda(() ->
                bestFit.ifPresent(sdk -> ProjectRootManager.getInstance(myProject).setProjectSdk(sdk))));
    }

    public boolean isArtemisExercise() {
        return myState != null;
    }

    public ExerciseView getCurrentView() {
        return myState.view;
    }
}

File: src/main/java/de/tum/www1/orion/exercise/registry/OrionProjectRegistryStateService.kt
package de.tum.www1.orion.exercise.registry

import com.intellij.notification.NotificationType
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.components.StoragePathMacros
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.projectRoots.ProjectJdkTable
import com.intellij.openapi.projectRoots.Sdk
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.vfs.VfsUtil
import com.jetbrains.python.sdk.PythonSdkUtil
import de.tum.www1.orion.dto.RepositoryType
import de.tum.www1.orion.enumeration.ExerciseView
import de.tum.www1.orion.enumeration.ProgrammingLanguage
import de.tum.www1.orion.exercise.OrionJavaTutorProjectCreator
import de.tum.www1.orion.ui.util.notify
import de.tum.www1.orion.util.JsonUtils
import de.tum.www1.orion.util.OrionFileUtils.getRoot
import de.tum.www1.orion.util.translate
import java.io.IOException

/**
 * Interface to persist data in IntelliJ's workspace.xml file using built-in features.
 * Stores all artemis specific data like exercise id and course id.
 *
 * For some reason this class needs to be implemented in Kotlin, otherwise the IDE crushes when
 * opened in instructor mode when running "build and test locally".
 */
@State(name = "orionRegistry", storages = [Storage(StoragePathMacros.WORKSPACE_FILE)])
class OrionProjectRegistryStateService(private val myProject: Project) :
    PersistentStateComponent<OrionProjectRegistryStateService.State?> {
    private var myState: State? = null

    /**
     * Renaming any field in state would be a breaking change and require the users to re-clone the repo.
     */
    data class State(
        var courseId: Long = 0,
        var courseTitle: String? = null,
        var exerciseId: Long = 0,
        var exerciseTitle: String? = null,
        var selectedRepository: RepositoryType = RepositoryType.ASSIGNMENT,
        var currentView: ExerciseView = ExerciseView.STUDENT,
        var language: ProgrammingLanguage = ProgrammingLanguage.JAVA,
        // For exercises opened as instructor
        var templateParticipationId: Long? = null,
        var solutionParticipationId: Long? = null,
        // For exercises opened as tutor
        var submissionId: Long? = null,
        var correctionRound: Long? = null,
        // Currently unused, Orion does not support test runs
        // The flag is transferred and stored nonetheless to allow quick extension
        var testRun: Boolean? = null,
        // For exam exercises
        var exerciseGroupId: Long? = null,
        var examId: Long? = null,
    )

    override fun getState(): State? {
        return myState
    }

    override fun loadState(state: State) {
        myState = state
    }

    /**
     * Store the information imported from the .artemisExercise.json received from the server.
     * The file is deleted afterwards.
     */
    fun importIfPending() {
        val pendingImportFile = VfsUtil.findRelativeFile(getRoot(myProject), ".artemisExercise.json")
        if (pendingImportFile != null) {
            try {
                val imported = JsonUtils.mapper().readValue(pendingImportFile.inputStream, ImportedExercise::class.java)
                val myState = State()
                imported.apply {
                    myState.courseId = courseId
                    myState.exerciseId = exerciseId
                    myState.courseTitle = courseTitle
                    myState.exerciseTitle = exerciseTitle
                    myState.language = language
                    myState.currentView = view
                    myState.templateParticipationId = templateParticipationId
                    myState.solutionParticipationId = solutionParticipationId
                    myState.exerciseGroupId = exerciseGroupId
                    myState.examId = examId
                }
                // operations specific to a view
                when (myState.currentView) {
                    ExerciseView.INSTRUCTOR -> {
                        myState.guessProjectSdk()
                        // init
                        myState.selectedRepository = RepositoryType.TEST
                    }
                    ExerciseView.TUTOR -> OrionJavaTutorProjectCreator.prepareProjectAfterImport(myProject)
                    else -> Unit
                }
                // warn if language not supported
                when (myState.language) {
                    ProgrammingLanguage.JAVA -> Unit
                    else -> myProject.notify(translate("orion.error.language.import.notSupported").format(myState.language.name))
                }
                loadState(myState)
                ApplicationManager.getApplication().invokeLater {
                    try {
                        WriteAction.run<IOException> { pendingImportFile.delete(this) }
                    } catch (e: IOException) {
                        log.error(e.message, e)
                    }
                }
            } catch (e: IOException) {
                log.error(e.message, e)
            }
        }
    }

    /**
     * Suggests SDK to use for the freshly opened project.
     */
    private fun State.guessProjectSdk() {
        val availableSdks: List<Sdk> = when (this.language) {
            ProgrammingLanguage.JAVA -> ProjectJdkTable.getInstance().allJdks.toList()
            ProgrammingLanguage.KOTLIN -> ProjectJdkTable.getInstance().allJdks.toList()
            ProgrammingLanguage.PYTHON -> PythonSdkUtil.getAllSdks()
            else -> return
        }
        if (availableSdks.isEmpty()) {
            myProject.notify(translate("orion.error.sdk.notAvailable"))
            return
        }
        availableSdks.maxWithOrNull(compareBy { sdk -> sdk.versionString })?.let {
            ApplicationManager.getApplication().invokeLater {
                try {
                    WriteAction.run<RuntimeException> {
                        ProjectRootManager.getInstance(myProject).projectSdk = it
                    }
                } catch (e: Throwable) {
                    myProject.notify(
                        translate("orion.error.exercise.sdkFailed").format(e.message),
                        NotificationType.WARNING
                    )
                }
            }
        }

    }

    val isArtemisExercise: Boolean
        get() = myState != null

    companion object {
        private val log = Logger.getInstance(
            OrionProjectRegistryStateService::class.java
        )
    }
}
